src/assets/scripts/client/aircraft/AircraftModel.js

```diff
@@ -848,15 +853,23 @@ export default class AircraftModel {
      * @return {boolean}
      */
     isBeyondTopOfDescent() {
-        const waypointModel = this.fms.nextHardAltitudeRestrictedWaypoint;
-
-        if (_isNil(waypointModel)) {
+        if (_isEmpty(this.fms.waypoints)) {
             return;
         }
 
-        const waypointAltitude = waypointModel.altitudeMaximum;
-        const waypointDistance = this.positionModel.distanceToPosition(waypointModel.positionModel);
-        const altitudeChange = waypointAltitude - this.altitude;
+        const currentAltitude = this.altitude;
+        const waypoints = this.fms.getAltitudeRestrictedWaypoints();
+        const waypointsWithRelevantCeiling = _filter(waypoints, (waypoint) => waypoint.hasMaximumAltitudeBelow(this.altitude));
+
+        const targetAltitude = _isEmpty(waypointsWithRelevantCeiling) ?
+              this.fms.getBottomAltitude() :
+              _head(waypointsWithRelevantCeiling).altitudeMaximum;
+        const targetPosition = _isEmpty(waypointsWithRelevantCeiling) ?
+              _last(this.fms.waypoints).positionModel :
+              _head(waypointsWithRelevantCeiling).positionModel;
```

Comment:
I will keep the verbosity with variables in mind, it's something I should definitely do more.

The example quoted here is not equivalent to the pull request code.  The intent is to check for the next waypoint with a Maximum altitude restriction which the aircraft needs to descend to in order to comply with, in the case there are no maximum altitude restrictions then the descent should be to the lowest minimum altitude restriction and should occur as late as possible to reach that altitude while still on the procedure.  This doesn't properly take into account the somewhat unlikely case where a procedure has a minimum restriction which must be met followed by a maximum restriction and the aircraft is unable to descend fast enough between the two to comply with both.  There is actually a couple of possible variations on that edge case which aren't quite properly dealt with.

As for the logic, this is one of the rare cases where a ternary may be the best option locally.  A guard statement isn't appropriate because we actually want to take one of two branches.  A bunch of the relevant state is in the current function so it's not an easy refactor into Fms.  The alternative to the ternary locally is an `if` statement but then we can't use the `const` keyword as it's block scoped so it becomes a `let` outside of the if which is overwritten only once on some procedures.

Reply:
Indeed the quoted example is not equivalent to the pull request code, because there are some issues with how that logic works out:

> in the case there are no maximum altitude restrictions then the descent should be to the lowest minimum altitude restriction 

this _is_ true, but also not really relevant to this particular method. If there are no further relevantly-restricted waypoints, then it doesn't matter what the bottom altitude is... we're past the point where there's any useful descent planning information, so the answer to the question of whether or not we are `past the top of descent` really can't be answered. And so, we can just return. Also, `_last(this.fms.waypoints)` is probably not the one you're looking for to match the bottom altitude anyway.

In general, for our purposes, overwriting a `let` declared variable would be highly preferable to a ternary, especially in cases like this where multiple ternaries are used contingent to a common condition.

> This doesn't properly take into account the somewhat unlikely case where a procedure has a minimum restriction which must be met followed by a maximum restriction and the aircraft is unable to descend fast enough between the two to comply with both. 

And this is true as well. That would be nice to have the aircraft advise the controller that they are unable to make the second restriction, but that's a shortcoming that isn't a big concern since most procedures will be designed to be conducive to performance of the aircraft frequently using it. Either way though, this is something that would be dealt with in the `_calcualateTargetAltitude....()` areas or maybe future methods where the descent planning is done for all waypoints, but is out of scope of this particular method, which should only be concerned with whether the FMS should remain in cruise mode or kick into descent mode.

Reply:
The ternary has been removed and replaced with `let` and re-assignment.

I've renamed the method to `shouldStartDescent` as that's really what it's being used to decide.  As such, it should start the descent in two scenarios:
1. There is a maximumAltitude restriction and the aircraft needs to descend to meet it.
2. A continuous descent along the planned flight path is required to intercept the expected approach.

The current implementation of the flight plan doesn't actually terminate at an airport or runway so I'm doing the next best thing and starting the descent at the point where a continuous descent would reach the minimum altitude at the end of the flight plan in case of scenario 2.  The Fms also doesn't know where the aircraft is so can't tell the distance remaining to any given waypoint, which means doing the direct distance to the last waypoint.